src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs

```diff
@@ -327,12 +327,28 @@ private bool IsCommitCharacter(char ch)
             var selectedItem = GetExternallyUsableCompletionItem(model.SelectedItem);
             if (selectedItem.IsBuilder)
             {
-                return char.IsLetterOrDigit(ch);
+                return char.IsLetterOrDigit(typedChar);
             }
 
-            var filterText = GetCurrentFilterText(model, selectedItem);
+            var snapshot = this.TextView.TextSnapshot;
+            var editSpan = model.GetCurrentSpanInSnapshot(model.GetSubjectBufferFilterSpanInViewBuffer(selectedItem.FilterSpan), snapshot);
+            var spanStart = (int)editSpan.Start;
+            var caretPosition = (int)GetCaretPointInViewBuffer();
+
+            var filterText = snapshot.GetText(Span.FromBounds(spanStart, caretPosition));
+            var isCommitCharacter = GetCompletionRules().IsCommitCharacter(selectedItem, typedChar, filterText);
+
+            // When a character, even appears to be a commit character, was typed at the begining of the filter span when it was not empty,
+            // it's most likely that the user just realized he/she forgot to do something, pressed Ctrl+Left and typed it.
+            // i.e. blah| -> |blah -> !|blah
+            // We should not commit the selected item in the above case.
+            // But be careful we still want completion in the following case: A a = new | -> A a = new (|
+            if (isCommitCharacter && caretPosition == spanStart + 1 && editSpan.Length > 1)
```

Comment:
IIRC @kayleh changed this, because in some asp.net cases the subject buffer may be disconnected and then re-connected to the view buffer in the course of a single session. :frowning: